libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -311,24 +328,49 @@ proc sendMsgSlow(p: PubSubPeer, msg: seq[byte]) {.async.} =
     debug "No send connection", p, msg = shortLog(msg)
     return
 
-  trace "sending encoded msg to peer", conn, encoded = shortLog(msg)
-  await sendMsgContinue(conn, conn.writeLp(msg))
-
-proc sendMsg(p: PubSubPeer, msg: seq[byte]): Future[void] =
+proc sendMsg(p: PubSubPeer, msg: seq[byte], saltedId: Option[SaltedId] = none(SaltedId)): Future[void] {.async.}=
+  if p.sendConn == nil or p.sendConn.closed():
+    await sendMsgSlow(p, msg)
+  
   if p.sendConn != nil and not p.sendConn.closed():
-    # Fast path that avoids copying msg (which happens for {.async.})
     let conn = p.sendConn
-
     trace "sending encoded msg to peer", conn, encoded = shortLog(msg)
-    let f = conn.writeLp(msg)
-    if not f.completed():
-      sendMsgContinue(conn, f)
-    else:
-      f
-  else:
-    sendMsgSlow(p, msg)
 
-proc sendEncoded*(p: PubSubPeer, msg: seq[byte], isHighPriority: bool): Future[void] =
+    if msg.len < 2000:
+      try:
+        let f = conn.writeLp(msg)
+        await f
+      except:
+        await conn.close()
+    else:
+      await p.semTxLimit[].acquire()
+      #We may have received DontWant for the message
+      if saltedId.isSome:
+        for heDontWant in p.heDontWants:
+          if saltedId.get in heDontWant:
+            p.semTxLimit[].release()
+            atomicInc(libp2p_gossipsub_staggerDontWantSave2)
+            return
+      try:
+        let f = conn.writeLp(msg)
+        let turns = (msg.len div 100_000) + 1
+        for i in 1..turns:
+          await f or sleepAsync(200.milliseconds)   #sleep time should be adaptive to the peer bandwidth
+          if not f.completed and saltedId.isSome:
+            for heDontWant in p.heDontWants:
+              if saltedId.get in heDontWant:
+                atomicInc(libp2p_gossipsub_staggerDontWantSave3)
+                break
+        if not f.completed:
+          await f.cancelAndWait()
+          #asyncSpawn sendMsgContinue(conn, f)
+        p.semTxLimit[].release()
+      
+      except LPStreamError as exc:
+        p.semTxLimit[].release()
+        await conn.close()
+    
+proc sendEncoded*(p: PubSubPeer, msg: seq[byte], isHighPriority: bool, saltedId: Option[SaltedId] = none(SaltedId)): Future[void] =
```

Comment:
why is saltedId optional?

Reply:
[heDontWants is storing saltedId](https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub/behavior.nim#L271)

Reply:
> why is saltedId optional?

same SendMsg for control and published messages, so for some messages we did not need saltedID

Reply:
Currently, we can improve this design as I believe high-prio are only control and msgs published by the local peer. It means no other peer can send an `idontwant` for them.  It means we can move the `idontwant` to the non-prio branch in this proc.

Reply:
In fact, If we were enqueueing those msgs and we didn't dequeue them fast enough, peers could receive them via peers we sent first. But as we always send them immediately, it seems this can't happen. Makes sense?

Reply:
> Currently, we can improve this design as I believe high-prio are only control and msgs published by the local peer. It means no other peer can send an `idontwant` for them. It means we can move the `idontwant` to the non-prio branch in this proc. Wdyt?

yes, we dont need to check hedontwants for the messages published by the local peer,,, but i guess IDontWant should be transmitted with the highest priority

Reply:
In fact, i believe, IDontWant should be issued before [this step](https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub.nim#L366). Because even if a message is not valid, we can react early to stop the spread of such message. And in case of valid messages, we get to learn early that our peers have received this message

Reply:
I meant that checking `idontwant` before sending a msg could be done only for msgs relayed at line 417 of this file.

Reply:
yes for relayed messages only